### PR TITLE
[INFRA] Use GitHub CI

### PR DESCRIPTION
**Pull request by eseiler**
Also:

* removes `set (CMAKE_CXX_STANDARD 17)`. this forces cpp17 when cpp20 is specified via `CMAKE_CXX_FLAGS`
* removes `set (CMAKE_CXX_STANDARD_REQUIRED ON)`
* updates SeqAn3 to 3.0.2

Follow-up PR:
* set the workflow ID in the cancel workflow (reason: see https://github.com/seqan/app-template/pull/34#issuecomment-742606822)
* Update the badge in the readme
* Set push to `master`, I commented it out such that it builds on my repository

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+on:
+  push:
+    branches:
+      # Push events to branches matching refs/heads/master
+      - 'master'
+  pull_request:
+
+env:
+  CMAKE_VERSION: 3.19.1
```
Comment (joergi-w) [inline .github/workflows/ci.yaml:11]: In the app template we currently require only CMake version ≥ 3.8.2. 
Should we adapt all CMake files or stay here with 3.8.2? Which features would we need from newer CMake?

Comment (eseiler) [reply .github/workflows/ci.yaml]: Changed it to 3.8.2 :)

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,8 +4,8 @@ cmake_minimum_required (VERSION 3.8)
 
 # Define the application name and version.
 project (fastq_to_fasta VERSION 1.0.0)
-set (CMAKE_CXX_STANDARD 17)
-set (CMAKE_CXX_STANDARD_REQUIRED ON)
+# set (CMAKE_CXX_STANDARD 17)
+# set (CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (eseiler) [inline CMakeLists.txt:8]: This seems wrong? We actually require cpp20, or cpp17 with concepts.

I have a problem on gcc10 cpp20 linux with it.
It appends and hence overwrite the `-std=c++20` flag I pass.
The resulting options are `-std=c++20 -std=c++1z`, hence it builds in cpp17 mode.

gcc10 on macos graciously adds `-fconcepts`, so it still builds.
On linux, the `-fconcepts` is not added, so it fails.

**Comment by eseiler** [comment]
It builds on my fork: https://github.com/eseiler/app-template/runs/1531370963

However, it doesn't show up in this repo. This may have to do something with this repo being a template repository (some rules for forks are different when it comes to template repositories). There is no documentation regarding template repos + github actions.

I'll create a template repo and check if it works as expected after merging.

**Comment by eseiler** [comment]
@smehringer and I reproduced the setup:

* This PR will not build on this repo, it will only build on my fork
* After merging, the merge commit will build on this repo
* Any PR after merging will build on this repo
* If a PR changes the workflow, the changes are already implied on the PR run

tl;dr:  this PR won't build, after that, everything works as usual

**Review by joergi-w — CHANGES_REQUESTED** [review]
Looks good, thank you!

Two questions: 
1. Why don't you use `apt-get` or `brew` to install doxygen and cmake?
2. Is it possible to "allow failures" (like in Travis)? I would prefer to use it for the documentation of apps. Not all app developers write exhaustive api documentation which covers 100% of each function parameter etc. and it should still set the overall green light that builds are passing (which is important for an app user).

**Comment by eseiler** [comment]
> Looks good, thank you!
> 
> Two questions:
> 
> 1. Why don't you use `apt-get` or `brew` to install doxygen and cmake?

So we can choose the version we want to test. It's also just copy/paste from existing workflows, so it's easier to maintain.
The main point would be that you do not get failed PRs just because you don't immediately support the most up to date version - e.g. doxygen kinda likes to break things with new releases.

> 2. Is it possible to "allow failures" (like in Travis)? I would prefer to use it for the documentation of apps. Not all app developers write exhaustive api documentation which covers 100% of each function parameter etc. and it should still set the overall green light that builds are passing (which is important for an app user).

I see your point and would argue that in this case you should just disable the run in the matrix instead of allowing it to fail.
Either you want your documentation to build and be free of errors or you do not. In the latter case, I would just not build it at all.
I can allow jobs to fail, but I would need to set the property for all jobs, i.e. I need to state for each job if it is allowed to fail or not (more than a one-liner).

**Review by joergi-w — APPROVED** [review]
Thanks for the convincing insights :)

**Review by smehringer — APPROVED** [review]
